core/src/main/java/org/apache/calcite/rel/rel2sql/PivotRelToSqlUtil.java

```diff
@@ -138,18 +162,35 @@ private SqlNodeList getAxisSqlNodes(Aggregate e) {
     return axesNodeList;
   }
 
-  private SqlNodeList getAggSqlNodes(Aggregate e, List<SqlNode> selectColumnList) {
+  private SqlNodeList getAxisNodeList(List<SqlNode> selectColumnList, boolean checksubquery) {
+
     final Set<SqlNode> selectList = new HashSet<>();
-    for (int i = 0; i < e.getAggCallList().size(); i++) {
-      int fieldIndex = e.getAggCallList().get(i).filterArg - (i + 2);
-      if (fieldIndex < 0) {
-        continue;
-      }
-      SqlNode aggCallSqlNode = selectColumnList.get(fieldIndex);
-      selectList.add(aggCallSqlNode);
 
+    SqlBasicCall pivotColumnAggregation =
+        (SqlBasicCall) selectColumnList.get(selectColumnList.size() - 1);
+
+    if (!checksubquery) {
+      SqlCase pivotColumnAggregationCaseCall =
+          (SqlCase) (
+              (SqlBasicCall) ((SqlBasicCall) selectColumnList
+              .get(selectColumnList.size() - 1))
+              .getOperandList().get(0)).getOperandList().get(0);
+      SqlBasicCall caseConditionCall =
+          (SqlBasicCall) pivotColumnAggregationCaseCall.getWhenOperands().get(0);
+      SqlIdentifier aggregateCol = caseConditionCall.operand(0);
+      selectList.add(aggregateCol);
+      return new SqlNodeList(selectList, pos);
     }
-    SqlNodeList aggList = new SqlNodeList(selectList, pos);
-    return aggList;
+
+    SqlBasicCall axisNodeList =
+        ((SqlBasicCall) pivotColumnAggregation.getOperandList().get(0)).operand(0);
+
+    if (axisNodeList.getOperator().kind == SqlKind.AS) {
+      selectList.add(axisNodeList.operand(1));
```

Comment:
u might be able to avoid the changes in SqlCharStringLiteral by doing some modifications.

Reply:
yes i have created Identifier for the axisColumn, so no need to handle it in unparsing